Makefile

```diff
@@ -1,26 +1,86 @@
-
-
 # For Go
 GO111MODULE = on
-GOFLAGS     = -mod=vendor
+GOOS := $(shell go env GOOS)
+GOARCH := $(shell go env GOARCH)
+GOFLAGS := -mod=vendor
 export GO111MODULE GOFLAGS
 
-all: test
+SUDO=sudo
+
+# Image URL to use all building/pushing image targets
+IMG ?= controller:latest
+# Produce CRDs with apiextensions.k8s.io/v1
+CRD_OPTIONS ?= "crd:crdVersions=v1"
+
+# Get the currently used golang install path (in GOPATH/bin, unless GOBIN is set)
+ifeq (,$(shell go env GOBIN))
+GOBIN=$(shell go env GOPATH)/bin
+else
+GOBIN=$(shell go env GOBIN)
+endif
 
+KUBEBUILDER_VERSION := 2.3.1
+CTRLTOOLS_VERSION := 0.2.9
+
+all: moco-controller
+
+# Run tests
 test:
+	cd /tmp; GO111MODULE=on GOFLAGS= go install github.com/cybozu/neco-containers/golang/analyzer/cmd/custom-checker
 	test -z "$$(gofmt -s -l . | grep -v '^vendor' | tee /dev/stderr)"
 	test -z "$$(golint $$(go list -mod=vendor ./... | grep -v /vendor/) | tee /dev/stderr)"
 	test -z "$$(nilerr ./... 2>&1 | tee /dev/stderr)"
-	test -z "$$(restrictpkg -packages=html/template,log ./... 2>&1 | tee /dev/stderr)"
-	go build -mod=vendor ./...
-	go test -mod=vendor -race -v ./...
-	go vet -mod=vendor ./...
+	test -z "$$(custom-checker -restrictpkg.packages=html/template,log $$(go list -tags='$(GOTAGS)' ./... | grep -v /vendor/ ) 2>&1 | tee /dev/stderr)"
 	ineffassign .
+	go build -mod=vendor ./...
+	go test -race -v -coverprofile cover.out ./...
+	go vet ./...
+	test -z "$$(go vet ./... | grep -v '^vendor' | tee /dev/stderr)"
+
+# Build moco-controller binary
+moco-controller: generate
+	go build -o bin/moco-controller ./cmd/moco-controller/main.go
+
+# Generate manifests e.g. CRD, RBAC etc.
+manifests: controller-gen
+	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=moco-controller-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+	rm -f deploy/crd.yaml
+	cp config/crd/bases/myso.cybozu.com_mysqlclusters.yaml deploy/crd.yaml
+
+# Generate code
+generate: controller-gen
+	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
+
+# find or download controller-gen
+# download controller-gen if necessary
+controller-gen:
+ifeq (, $(shell which controller-gen))
+	@{ \
+	set -e ;\
+	CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
+	cd $$CONTROLLER_GEN_TMP_DIR ;\
+	go mod init tmp ;\
+	go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.5 ;\
```

Comment:
Why don't you use `CTRLTOOLS_VERSION` as previously defined?
Or if installing controller-gen is a matter of `make setup`, remove this target altogether.

Reply:
Removed `controller-gen` target.